### PR TITLE
Arm backend: Make composable_quantizer default

### PR DESCRIPTION
**Pull request by AdrianLundell**
A few fixes needed:
- Add new ops added since initial upstream of composable_quantizer
- Add while-op quantize fix from 3be45468b to TosaQuantizerV2
- Add fixed_qparams fix from fb90480a261a72cae to TosaQuantizerV2
- Update some tests to mirror new behaviours
- Update quanitzer_tutorial to not be WIP
- Remove hardsigmoid from FUSED_ACTIVATION_OPS
- Explicitly check that weights and biases are input args to conv/ linear ops. The assumption that wights and biases are the only parameters of networks does not hold for real models.

cc @digantdesai @freddan80 @per @zingo @oscarandersson8218 @mansnils @Sebastian-Larsson @robell @rascani

### CODE REVIEW

**Comment by rascani** [comment]
Kicking off an internal test run.

**Comment by rascani** [comment]
It looks like the new quantizer breaks `examples/models/llama/tests/test_export_llama_lib.py`, because of asserts on quantizer.module_type_config, which is not implemented by the new quantizer. I think we should probably just remove those asserts in the tests unless you want to expose the module type config in the new quantizer.

https://github.com/pytorch/executorch/blob/main/examples/models/llama/tests/test_export_llama_lib.py#L121

File: backends/arm/quantizer/arm_quantizer_utils.py [hunk 1]
```diff
@@ -275,75 +287,59 @@ def get_quantizer_info(self):
...
+        # Ensure that the node is used as a bias by all users
+        for user_node in node.users:
+            if user_node.target not in self.PARAMETER_TARGETS:
+                return False
+
+            args = list(user_node.args)
+            if not (len(args) > 2 and node == args[2]):
+                return False
+
+        return True
```
[the remark this answers is not in the document]
Comment (AdrianLundell) [reply backends/arm/quantizer/arm_quantizer_utils.py:320]: This is not a regression, the embedding is marked for quantization and then decomposed into a index_select before real quantization, which is what is happening currently. Will add this to the backlog to ensure we don't want weight qspecs here though.

**Comment by rascani** [comment]
Kicking off another internal run.

**Comment by digantdesai** [comment]
FYI @3l1 - not sure if you saw this.

**Comment by digantdesai** [comment]
From @rascani 's internal CI runs,
```
executorch/backends/arm/quantizer/arm_quantizer.py", line 522, in global_config
    raise NotImplementedError(
NotImplementedError: Composable quantizer does not allow setting global_config directly. Please use set_global() instead.
```

These should be forward fixed.

**Comment by rascani** [comment]
> From @rascani 's internal CI runs,
> 
> ```
> executorch/backends/arm/quantizer/arm_quantizer.py", line 522, in global_config
>     raise NotImplementedError(
> NotImplementedError: Composable quantizer does not allow setting global_config directly. Please use set_global() instead.
> ```
> 
> These should be forward fixed.

I already have these fixed in D106539874. I'll land it ahead of time so the default switch can be a no-op.

**Comment by rascani** [comment]
Just to update, there's still some broken internal tests in switching to the composable quantizer. I'm working my way through them, and will approve once its clean.

**Comment by AdrianLundell** [comment]
@rascani Is it time to rebase this now and try it again or are there more fixes required?

**Comment by rascani** [comment]
> @rascani Is it time to rebase this now and try it again or are there more fixes required?

Sorry for the delay @AdrianLundell. This proved tricky to test out due to the API changes (the setter removal in particular). I've exported PR #20291 with a fix for an edge case where I/O was getting left in float. The fix feels a little hacky to me, but I also added an OSS test case for the edge case. So feel free to reuse the test with a better fix.

I have gotten a green internal test run with a stack of:
* PR #20291
* Internal switch to new V2 on users that would break with the new API
* Internal updated version of this PR switching the default

**Comment by AdrianLundell** [comment]
@rascani Which setter-removal do you mean? I don't think setting the config from a dict has been the intended API, setting it using the setter functions should be intact.

**Comment by rascani** [comment]
> @rascani Which setter-removal do you mean? I don't think setting the config from a dict has been the intended API, setting it using the setter functions should be intact.

We hit many of the NotImplementedError exceptions in this block: 

https://github.com/pytorch/executorch/blob/e28ef13dd7452fc9bc935f84d553b9f0d59b6877/backends/arm/quantizer/arm_quantizer.py#L539-L603

**Comment by AdrianLundell** [comment]
I would recommend replacing accessing e.g. `module_name_config` directly with using the official `set_module_name` setter API in that case, does that resolve the issue?

**Comment by AdrianLundell** [comment]
I realize we could also make a setter which takes the dict and calls `set_module_name` on all entries to have no breakage, would that help?

**Comment by AdrianLundell** [comment]
I had to take a break from this one, but now you should be able to set quantization specs from dicts as you are used to @rascani.

**Comment by AdrianLundell** [comment]
Look like the quantizers have diverged again, I will fix in a separate commit.

**Comment by digantdesai** [comment]
Still blocked on us, sorry.

**Review by digantdesai — APPROVED** [review]
Review automatically exported from Phabricator review in Meta.

**Comment by zingo** [comment]
Yeah!

**Comment by digantdesai** [comment]
> Review automatically exported from Phabricator review in Meta.

sorry error on my part. Let me revert.

**Comment by AdrianLundell** [comment]
I see, when would you say this will be unblocked and can I make any updates to make it easier? @digantdesai

**Comment by rascani** [comment]
> I see, when would you say this will be unblocked and can I make any updates to make it easier? @digantdesai

We had a bit of a race condition with an internal change. Barring no surprises, I am hoping to reland this today.

**Comment by rascani** [comment]
> > I see, when would you say this will be unblocked and can I make any updates to make it easier? @digantdesai
> 
> We had a bit of a race condition with an internal change. Barring no surprises, I am hoping to reland this today.

🤞 I think we should be good now.
